xds/src/main/java/io/grpc/xds/XdsDependencyManager.java

```diff
@@ -0,0 +1,597 @@
+    this.dataPlaneAuthority = checkNotNull(dataPlaneAuthority, "dataPlaneAuthority");
+
+    // start the ball rolling
+    addWatcher(new LdsWatcher(listenerName));
+  }
+
+  @Override
+  public ClusterSubscription subscribeToCluster(String clusterName) {
+
+    checkNotNull(clusterName, "clusterName");
+    ClusterSubscription subscription = new ClusterSubscription(clusterName);
+
+    Set<ClusterSubscription> localSubscriptions =
+        clusterSubscriptions.computeIfAbsent(clusterName, k -> new HashSet<>());
+    localSubscriptions.add(subscription);
+    addWatcher(new CdsWatcher(clusterName));
+
+    return subscription;
+  }
+
+  private boolean hasWatcher(XdsResourceType<?> type, String resourceName) {
+    TypeWatchers<?> typeWatchers = resourceWatchers.get(type);
+    return typeWatchers != null && typeWatchers.watchers.containsKey(resourceName);
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T extends ResourceUpdate> void addWatcher(XdsWatcherBase<T> watcher) {
+    XdsResourceType<T> type = watcher.type;
+    String resourceName = watcher.resourceName;
+
+    this.syncContext.execute(() -> {
+      TypeWatchers<T> typeWatchers = (TypeWatchers<T>)resourceWatchers.get(type);
+      if (typeWatchers == null) {
+        typeWatchers = new TypeWatchers<>(type);
+        resourceWatchers.put(type, typeWatchers);
+      }
+
+      typeWatchers.add(resourceName, watcher);
+      xdsClient.watchXdsResource(type, resourceName, watcher);
+    });
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T extends ResourceUpdate> void cancelWatcher(XdsWatcherBase<T> watcher) {
+    if (watcher == null) {
+      return;
+    }
+
+    XdsResourceType<T> type = watcher.type;
+    String resourceName = watcher.resourceName;
+
+    this.syncContext.execute(() -> {
+      TypeWatchers<T> typeWatchers = (TypeWatchers<T>)resourceWatchers.get(type);
+      if (typeWatchers == null) {
+        logger.log(DEBUG, "Trying to cancel watcher {0}, but type not watched", watcher);
+        return;
+      }
+
+      typeWatchers.watchers.remove(resourceName);
+      xdsClient.cancelXdsResourceWatch(type, resourceName, watcher);
+    });
+
+  }
+
+  public void shutdown() {
+    for (TypeWatchers<?> watchers : resourceWatchers.values()) {
+      shutdownWatchersForType(watchers);
+    }
+    resourceWatchers.clear();
+  }
+
+  private <T extends ResourceUpdate> void shutdownWatchersForType(TypeWatchers<T> watchers) {
+    for (Map.Entry<String, XdsWatcherBase<T>> watcherEntry : watchers.watchers.entrySet()) {
+      xdsClient.cancelXdsResourceWatch(watchers.resourceType, watcherEntry.getKey(),
+          watcherEntry.getValue());
+    }
+  }
+
+  private void releaseSubscription(ClusterSubscription subscription) {
+    checkNotNull(subscription, "subscription");
+    String clusterName = subscription.getClusterName();
+    Set<ClusterSubscription> subscriptions = clusterSubscriptions.get(clusterName);
+    if (subscriptions == null) {
```

Comment:
`|| !subscriptions.remove(subscription)`, since that is a common "already released" case?